server/internal/usecase/interactor/storytelling.go

```diff
@@ -468,17 +474,182 @@ func (i *Storytelling) RemovePageLayer(ctx context.Context, inp interfaces.PageL
 	return story, page, nil
 }
 
-func (i *Storytelling) CreateBlock(ctx context.Context, param interfaces.CreateBlockParam, operator *usecase.Operator) (*storytelling.Page, *storytelling.Block, error) {
-	// TODO implement me
-	panic("implement me")
+func (i *Storytelling) CreateBlock(ctx context.Context, inp interfaces.CreateBlockParam, op *usecase.Operator) (*storytelling.Story, *storytelling.Page, *storytelling.Block, int, error) {
+	tx, err := i.transaction.Begin(ctx)
+	if err != nil {
+		return nil, nil, nil, -1, err
+	}
+
+	ctx = tx.Context()
+	defer func() {
+		if err2 := tx.End(ctx); err == nil && err2 != nil {
+			err = err2
+		}
+	}()
+
+	story, err := i.storytellingRepo.FindByID(ctx, inp.StoryID)
+	if err != nil {
+		return nil, nil, nil, -1, err
+	}
+	if err := i.CanWriteScene(story.Scene(), op); err != nil {
+		return nil, nil, nil, -1, err
+	}
+
+	_, extension, err := i.getPlugin(ctx, &inp.PluginID, &inp.ExtensionID)
+	if err != nil {
+		return nil, nil, nil, -1, err
+	}
+	if extension.Type() != plugin.ExtensionTypeStoryBlock {
+		return nil, nil, nil, -1, interfaces.ErrExtensionTypeMustBeStoryBlock
+	}
+
+	prop, err := property.New().NewID().Schema(extension.Schema()).Scene(story.Scene()).Build()
+	if err != nil {
+		return nil, nil, nil, -1, err
+	}
+
+	block, err := storytelling.NewBlock().
+		NewID().
+		Plugin(inp.PluginID).
+		Extension(inp.ExtensionID).
+		Property(prop.ID()).
+		Build()
+	if err != nil {
+		return nil, nil, nil, -1, err
+	}
+
+	index := -1
+	if inp.Index != nil {
+		index = *inp.Index
+	}
+
+	page := story.Pages().Page(inp.PageID)
+	if page == nil {
+		return nil, nil, nil, -1, interfaces.ErrPageNotFound
+	}
+
+	page.AddBlock(block, index)
+
+	err = i.propertyRepo.Save(ctx, prop)
+	if err != nil {
+		return nil, nil, nil, -1, err
+	}
+
+	err = i.storytellingRepo.Save(ctx, *story)
+	if err != nil {
+		return nil, nil, nil, -1, err
+	}
+
+	tx.Commit()
+	return story, page, block, 1, err
 }
```

Comment:
Looks like you can separate the transaction-related functions and make them as a higher-order function which accepts the function to execute as an argument.

Something like
```
func (i *Storytelling) executeInTransaction(ctx context.Context, toExecute func(context.Context) error) error {
	tx, err := i.transaction.Begin(ctx)
	if err != nil {
		return err
	}
	ctx = tx.Context()
	defer func() {
		if err2 := tx.End(ctx); err == nil && err2 != nil {
			err = err2
		}
	}()
	err = toExecute(ctx)
	if err == nil {
		tx.Commit()
	}
	return err
}
```
You can refactor other functions (RemoveBlock, MoveBlock) in a similar way, extracting out common parts to helper methods or functions and using the executeInTransaction higher-order function to encapsulate the transaction handling code.

Reply:
Will work on this on different PR